File: router/workers/content.py
"""Content path — keyword and language matching on posts, raising notifications.

Latency-sensitive by nature: a notification that arrives ten minutes late is
worth much less than one that arrives now. That is why this is the one worker
pinned at ``minReplicas: 1`` instead of scaling from zero — a cold start on the
notification path is a user-visible delay, whereas a cold start on an
aggregation path costs nothing but a slightly later alert.

**On handling firehose content.** The brief warns that raw post text can contain
anything. This handler therefore treats post text as something to match against,
never something to emit: alerts carry the DID, the record key, and *which*
keyword matched — never the text itself. Logs are the easiest place to
accidentally build a permanent, widely-readable copy of exactly the content you
were told to be careful with.
"""

from __future__ import annotations

from typing import Any

from ..observability import get_logger
from ..routing import OP_DELETE
from .runner import WorkerContext

log = get_logger("content")


class ContentHandler:
    name = "content"

    def __init__(self) -> None:
        self.ctx: WorkerContext | None = None
        self.keywords: list[str] = []
        self.languages: set[str] = set()

    async def setup(self, ctx: WorkerContext) -> None:
        self.ctx = ctx
        self.keywords = [k.lower() for k in ctx.settings.keywords]
        self.languages = {lang.lower() for lang in ctx.settings.languages}
        log.info(
            "content handler ready",
            extra={"context": {"keywords": len(self.keywords),
                               "languages": sorted(self.languages) or "any"}},
        )

    async def handle(self, event: dict[str, Any], headers: dict[str, str]) -> None:
        assert self.ctx is not None
        commit = event.get("commit") or {}

        # Deletes are routed to the retraction path, so seeing one here means the
        # routing table and this worker disagree. Loud, not silent.
        if commit.get("operation") == OP_DELETE:
            log.warning("delete reached content path; check routing")
            return

        record = commit.get("record") or {}
        text = record.get("text")
        if not isinstance(text, str) or not text:
            return

        langs = {str(x).lower() for x in (record.get("langs") or []) if x}
        if self.languages and not (langs & self.languages):
            return

        haystack = text.lower()
        matched = [kw for kw in self.keywords if kw in haystack]
        if not matched:
            return

        self.ctx.alert(
            self.name,
            "keyword-match",
            did=event.get("did"),
            rkey=commit.get("rkey"),
            collection=commit.get("collection"),
            matched=matched,
            langs=sorted(langs),
            # Deliberately: the length of the post, not the post.
            text_len=len(text),
        )

File: internal/handler/content.go
package handler

import (
	"context"
	"regexp"
	"strings"

	"github.com/shalevsharfsh/jetstream-router/internal/event"
)

// Content matches posts against a configured keyword and language set.
//
// Stateless, so it needs no sharding. Latency-sensitive by nature: a
// notification that arrives ten minutes late is worth much less than one that
// arrives now.
type Content struct {
	sink     Sink
	patterns []keyword
	langs    map[string]bool
}

type keyword struct {
	word string
	re   *regexp.Regexp
}

// NewContent compiles the keyword set once at startup.
//
// The boundary is a lookaround pair, not \b. \b is a *transition* between a word
// and a non-word character, so `\bc\+\+\b` can never match — the character after
// '+' would have to be a word character. The lookaround form asks the question
// actually meant, "not glued to a word", and handles c++, .NET and ai alike.
//
// Plain substring matching was the obvious first implementation and it is wrong
// in a way that only shows up against real data: with "ai" in the list it fires
// on said, again, email and Dubai. On a notification path that is worse than not
// matching at all, because it trains whoever receives the alerts to ignore them.
//
// Known limitation: word characters are undefined for scripts that do not
// delimit words with spaces. The language filter is the honest mitigation.
func NewContent(sink Sink, keywords, langs []string) *Content {
	c := &Content{sink: sink, langs: map[string]bool{}}
	for _, l := range lower(langs) {
		c.langs[l] = true
	}
	for _, w := range keywords {
		w = strings.TrimSpace(w)
		if w == "" {
			continue
		}
		re := regexp.MustCompile(`(?i)(?:^|[^\p{L}\p{N}_])` +
			regexp.QuoteMeta(w) + `(?:[^\p{L}\p{N}_]|$)`)
		c.patterns = append(c.patterns, keyword{word: strings.ToLower(w), re: re})
	}
	return c
}

// Handle matches one post.
func (c *Content) Handle(_ context.Context, _ int, ev event.Event) error {
	// Deletes route to retraction, so one arriving here means the routing table
	// and this handler disagree.
	if ev.Key.Operation == event.OpDelete {
		return nil
	}

	var rec postRecord
	if !decode(ev.Record, &rec) || rec.Text == "" {
		return nil
	}

	if len(c.langs) > 0 {
		ok := false
		for _, l := range lower(rec.Langs) {
			if c.langs[l] {
				ok = true
				break
			}
		}
		if !ok {
			return nil
		}
	}

	var hit []string
	for _, k := range c.patterns {
		if k.re.MatchString(rec.Text) {
			hit = append(hit, k.word)
		}
	}
	if len(hit) == 0 {
		return nil
	}

	// The text itself is deliberately absent: its length, not its content.
	c.sink.Alert("info", "content", "keyword matched",
		"did", ev.DID, "rkey", ev.RKey,
		"matched", hit, "langs", rec.Langs, "text_len", len(rec.Text))
	return nil
}
